lib/teleterm/services/userpreferences/userpreferences.go

```diff
@@ -54,21 +68,38 @@ func Update(ctx context.Context, rootClient Client, leafClient Client, newPrefer
 	// the fields that change.
 	// Calling `UpsertUserPreferences` with only the modified values would reset
 	// the rest of the preferences.
-	rootPreferencesResponse, err := rootClient.GetUserPreferences(ctx, &userpreferencesv1.GetUserPreferencesRequest{})
-	if err != nil {
+	getGroup, getGroupCtx := errgroup.WithContext(ctx)
+	var rootPreferencesResponse *userpreferencesv1.GetUserPreferencesResponse
+	var leafPreferencesResponse *userpreferencesv1.GetUserPreferencesResponse
+
+	getGroup.Go(func() error {
+		res, err := rootClient.GetUserPreferences(getGroupCtx, &userpreferencesv1.GetUserPreferencesRequest{})
+		rootPreferencesResponse = res
+		return trace.Wrap(err)
+	})
+
+	if leafClient != nil {
+		getGroup.Go(func() error {
+			res, err := leafClient.GetUserPreferences(getGroupCtx, &userpreferencesv1.GetUserPreferencesRequest{})
+			leafPreferencesResponse = res
+			return trace.Wrap(err)
+		})
+	}
+
+	if err := getGroup.Wait(); err != nil {
 		return nil, trace.Wrap(err)
 	}
-	rootPreferences := rootPreferencesResponse.GetPreferences()
 
+	rootPreferences := rootPreferencesResponse.GetPreferences()
 	var leafPreferences *userpreferencesv1.UserPreferences
-	if leafClient != nil {
-		response, err := leafClient.GetUserPreferences(ctx, &userpreferencesv1.GetUserPreferencesRequest{})
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		leafPreferences = response.GetPreferences()
+	if leafPreferencesResponse != nil {
+		leafPreferences = leafPreferencesResponse.GetPreferences()
 	}
 
+	// We do not use errgroup.WithContext since we don't want to cancel
+	// the other request when one of them fails.
+	upsertGroup := errgroup.Group{}
```

Comment:
So we're fine with the leaf cluster upsert succeeding while the root cluster upsert fails? If so, this comment could explain why that is fine and why we don't need to do this sequentially – there's no way we're going to remember that in six months. ;)

Reply:
I keep forgetting that this is how it works. It'd be a good idea to add a comment to `userpreferences.Update` that summarizes how the function works when called for a root cluster vs a leaf cluster.

Reply:
Good idea, added.